DRILL-6385: Support JPPD feature

weijietong opened the pull request:
@amansinha100 @parthchandra The join predicate push down feature is basically ready to review. Please give some review advices, thanks.

weijietong commented:
ping @amansinha100 @parthchandra or someone else would be pleased to review this PR.

amansinha100 commented:
@weijietong thanks for the PR.  Yes, this is high on my list to review.  I am tied up until tomorrow but will review during the weekend.

chunhui-shi commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/config/HashJoinPOP.java:
```diff
@@ -32,35 +33,53 @@
 public class HashJoinPOP extends AbstractJoinPop {
   static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(HashJoinPOP.class);
 
+  private RuntimeFilterDef runtimeFilterDef;
```
Who is responsible for constructing this runtimeFilterDef?

chunhui-shi replied:
After physical plan is generated, RuntimeFilterManager is is responsible for constructing this runtimeFilterDef. So I guess your estimation of the cost sending bloom filter is trivial.

weijietong replied:
RuntimeFilterDef was created after the physical plan was generated. Constructing the BloomFilter at HashJoin node is a heavy work， to a light weight join query (i.e. two small tables), it's not applicable to construct a bloom filter.So I left some TODO work at the RuntimeFilterManager.quelifiedHolders method to value whether applicable to apply the RuntimeFilter according to the NDV from both sides. The NDV statistics is not supported now. I will fire another work about the Metadata to support it.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/ScanBatch.java:
```diff
@@ -226,6 +244,96 @@ public IterOutcome next() {
+        ValueVector to = backUpContainer.getValueVector(i).getValueVector();
+        to.setInitialCapacity(svIndex);
+        for (int r = 0; r < svIndex; r++) {
+          to.copyEntry(r, from, selectionVector2.getIndex(r));
```
To summarize the way you have implemented this part:  suppose the original ScanBatch contained 100 rows and 10 of them qualified the bloom filter, you create an SV2 of size 10, set the original qualifying row's index in the SV2, then for each ValueVector you copy the qualifying row's data into the backupContainer, followed by exchanging it with the current output container.   A couple of thoughts about this: 
  - Why not produce the SV2 in the output batch and let the downstream operator handle it ? Quite often there may be a Filter operator above the Scan  which would be applying other filters (i.e not the run-time filters) and it can combine its own SV2 with the SV2 produced by the Scan.  That way you avoid the extra copying and let Filter handle it (note that Filter does code-gen so it is more efficient way to handle bulk filtering). 
 - There is clearly trade-offs with the extra copy approach : it depends on selectivity, i.e how many rows actually get eliminated by the run-time filter.  I suppose you have mentioned this as a TODO depending on the NDV statistics ?

weijietong replied:
@aman thanks for other valuable reviews ,will soon update that. To this point, I initially tend to output the SV2. But to fellow reasons, I give up: 

* Not all the possible operators support the SelectionModel. If users’ rule pushed down the filter conditions,the output SV2 maybe not processed by the not supported operators(i.e. it’s not definitive  to have a operator which supports the SelectionModel above the Scan node).

* The BatchSchema’s SelectionModel also becomes a runtime var.This will also affect the upper filter node’s code-gen logic to dynamically generate fresh filter codes to the Scan’s SV2.

I agree that the memory copy cost will be less if the above filter node can filter more rows over the Scan’s SV2.So what your opinion about this?

amansinha100 replied:
Hi @weijietong ,  the physical plan operators are supposed to implement 2 APIs:  `getSupportedEncodings()` API (see [1]) which indicates whether or not it accepts an SelectionVector2 (other options are SV4 and NONE) and the corresponding `getEncoding()` API.  If an operator does not accept an SV2 or SV4, the planner will insert a SelectionVectorRemover just below that node which essentially does the copying of qualified rows.  Note that the SVRemover is implemented by RemovingRecordBatch [2] which uses a `StraightCopier` whenever the child has NONE, so in that case it just does a simple transfer instead of copy.  

In the case of Filter,  it can accept both NONE and SV2 because it is possible in some cases to have a filter on top of another filter (with some intermediate non-blocking operator).  

Let me know if this makes sense.   The main reason I am proposing this is the Filter-Scan is a very common pattern and we would want to minimize the copy overhead as much as possible.  Unfortunately, I am tied up with some other work otherwise I could make the changes on top of your branch to experiment. 

[1] https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/Prel.java#L51
[2] https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/svremover/RemovingRecordBatch.java#L57

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java:
```diff
@@ -696,6 +780,18 @@ public void executeBuildPhase() throws SchemaChangeException {
+          for (BloomFilter bloomFilter : bloomFilters) {
+            for (int ind = 0; ind < currentRecordCount; ind++) {
+              long hashCode = hash64.hash64Code(ind, 0, condFieldIndex);
+              bloomFilter.insert(hashCode);
```
Currently, the hash join relies on memory calculations (both build and probe sides) to do accounting for spilling purposes.  The bloom filter memory use should also be included in that calculation, although it would fine if you create an enhancement JIRA and address it separately.

weijietong replied:
I will enhance this at another JIRA.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/config/HashJoinPOP.java:
```diff
@@ -32,35 +33,53 @@
+
 
 @JsonTypeName("hash-join")
+@JsonIgnoreProperties(ignoreUnknown = true)
```
The primary reason is to backward compatible. PhysicalPlans without `runtimeFilterDef` property (maybe not support RuntimeFilter) can still work out. That is the `ignoreUnknown=true` usage.  The `@JsonIgnore` is to not output the property content as the final json. That's not our target.

weijietong commented:
@amansinha100  The scan node's memory copy logic has removed. Thanks for the knowledge of `SelectionVectorPrelVisitor.addSelectionRemoverWhereNecessary`. I appreciate what you payed out. Thanks!

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterManager.java:
```diff
@@ -0,0 +1,755 @@
+    private List<CoordinationProtos.DrillbitEndpoint> probeSideScanEndpoints;
+
+    //probe side's children have blocked nodes.
+    private boolean meetBlockedOperator;
```
Better name would be 'probeHasBlockingOperator'

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterManager.java:
```diff
@@ -0,0 +1,755 @@
+    //probe side's children have blocked nodes.
+    private boolean meetBlockedOperator;
+
+    //whethe this join operator is a partitioned HashJoin or broadcast HashJoin
```
In addition to broadcast and partitioned hash join, there's a third state also which is non-distributed hash join (this depends on slice_target, estimated row count etc.).  Is the JPPD not applied for such cases ?

weijietong replied:
You mean there's only one HashJoin execution node? To that case , I guess JPPD is not necessary, since non-distributed one means less data, it's not cost efficient to generate a BF.

amansinha100 replied:
Yeah, it would be fine to not do JPPD for that case .. we can document the expected behavior.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterManager.java:
```diff
@@ -0,0 +1,755 @@
+    //whethe this join operator is a partitioned HashJoin or broadcast HashJoin
+    private boolean meetBroadcastExchange;
+
+    private boolean meetPartitionedExchange;
```
Better names would be `encounteredPartitionExchange`,  `encounteredBroadcastExchange`.   But do you really need these variables declared in the HashJoinOpHolder class since these are temporary states encountered during the traversal of the operator tree and you are anyways having another class variable `isBroadcastHashJoin`.

amansinha100 commented:
@weijietong I have couple of design questions that are better discussed in the JIRA, so I will add them there.. pls take a look.  Thanks.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/ScanPrel.java:
```diff
@@ -151,12 +151,12 @@ public RelOptCost computeSelfCost(final RelOptPlanner planner, RelMetadataQuery
   @Override
   public SelectionVectorMode[] getSupportedEncodings() {
-    return SelectionVectorMode.DEFAULT;
+    return SelectionVectorMode.NONE_AND_TWO;
```
Yes, I will update that.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/ScanPrel.java:
```diff
@@ -151,12 +151,12 @@ public RelOptCost computeSelfCost(final RelOptPlanner planner, RelMetadataQuery
   @Override
   public SelectionVectorMode getEncoding() {
-    return SelectionVectorMode.NONE;
+    return SelectionVectorMode.TWO_BYTE;
```
What do you mean by `NONE_AND_TWO ` ,since the return is `SelectionVectorMode` not an array. If JPPD is not applicable, to `ScanBatch` ,the output container's build schema would always be `SelectionVectorMode.NONE` , and will not generate SV2.

amansinha100 replied:
Makes sense.  I was thinking of the other API which returns an array.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/options/OptionValue.java:
```diff
@@ -76,7 +77,7 @@ public boolean inScopeOf(OptionScope scope) {
 
   public enum Kind {
-    BOOLEAN, LONG, STRING, DOUBLE
+    BOOLEAN, LONG, STRING, DOUBLE, INTEGER
```
The reason to add an INTEGER option is that the largest BF's byte size is supposed to be Integer.MAX_VALUE which is limited by DrillBuf's size.  So INTEGER option would be precisely.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/WorkManager.java:
```diff
@@ -375,6 +378,30 @@ protected void cleanup() {
+      //to foreman
+      if (toForeman) {
+        Foreman foreman = queries.get(queryId);
+        if (foreman != null) {
```
Yes, it's supposed to handle the cases: the foreman has crashed or the whole query has completed ahead of time.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterManager.java:
```diff
@@ -0,0 +1,755 @@
+    List<BloomFilterDef> bloomFilterDefs = new ArrayList<>();
+    for (JoinCondition joinCondition : joinConditions) {
+      LogicalExpression leftProExp = joinCondition.getLeft();
+      FieldReferenceFinder fieldReferenceFinder = new FieldReferenceFinder();
```
You are right, will update that to consider a join condition with function.

amansinha100 commented:
@weijietong thanks for making these changes.  I will do one final round of review hopefully soon and let you know.

weijietong commented:
ping @amansinha100

amansinha100 commented:
Hey @weijietong,  sorry for the delay; I will send you my final comments by tomorrow.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/BloomFilterCreator.java:
```diff
@@ -0,0 +1,41 @@
+import io.netty.buffer.DrillBuf;
+import org.apache.drill.exec.memory.BufferAllocator;
+
+public class BloomFilterCreator {
```
Could you pls add a description either here or another appropriate place about how the BF creation works for a generalized multi-column join.. e.g  t1.a1 = t2.a2 AND t1.b1 = t2.b2.   In this case, assume t2 is the build side.  My understanding is as follows: 
  - create a compound hash value using both build side columns: hash(a2, hash(b2))   (this may not be the exact way it is being created but same general idea).
  - add this to the BF, setting appropriate bits to 1
  - on probe side, do the same for a1, b1 and check BF for either false/possibly true

Let me know if this is correct.   This also implies that the size of BF is independent of the number of columns being joined.

weijietong replied:
Current implementation is one bloom filter one join column.  To your example, multi-column join , will generate two bloom filters.  The reason to this implementation is to achieve one vector column memory access by one  hash computation. But the Murmur hash 's complex computation ate up the pipeline performance , the result performance does not so good. So I will change it to your assumes implementation.

amansinha100 replied:
Hmm...if the 2 columns create separate bloom filters, where does the ANDing (intersection) of the two BFs happen ?  In the current implementation there's an ORing of the BFs but I did not see ANDing (maybe I missed, so let me know).  As far as the memory usage is concerned, we can continue down this path and evaluate both CPU cost and memory usage for multi-column joins where the cardinality of the columns ranges from low to high.

weijietong replied:
There's no intersection of the two separate BFs. The intersection meaning is implemented indirectly at the probe side `ScanBatch`. The separated BFs filter the probe side columns independently and set a BitSet (which indicates the possible rows matched the join condition ) together. This part of codes is at the `ScanBatch`'s `applyRuntimeFilter` and `computeBitSet` methods.

weijietong replied:
Btw, if there's too much multi-column joins , we can add an option to limit the generated  BloomFilter numbers. That's to say there's no need to generate one BF one joint column. We can choose to abandon generating some BFs to save the CPU cost and memory usage.

amansinha100 replied:
It would be good to add javadoc with this description of multi-column join and how the implicit intersection is done later during BF application in the Scan rather than at the time of BF creation.   This is different from normal hash join where a single hash value is created for multi-column during creation and probing.

weijietong replied:
This description will be add at the `ScanBatch`'s `applyRuntimeFilter `.

amansinha100 commented on exec/java-exec/src/main/resources/drill-module.conf:
```diff
@@ -455,6 +455,8 @@ drill.exec.options: {
     exec.hashjoin.num_partitions: 32,
     exec.hashjoin.num_rows_in_batch: 1024,
     exec.hashjoin.max_batches_in_memory: 0,
+    exec.hashjoin.enable.runtime_filter: true,
```
I would suggest to keep the default value as false until the feature goes through performance and functional testing.

weijietong replied:
agree

amansinha100 commented:
@weijietong  thanks for your patience.  I have submitted a final round of review comments.   Once you address them we can get the feature merged into 1.15.0 such that it gives it enough time to go through the normal functional and performance testing.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterManager.java:
```diff
@@ -0,0 +1,586 @@
+   * @param leftRelNode left RelNode of a BiRel or the SingleRel
+   * @return a left scan Prel which contains the left join condition name or null
+   */
+  private static ScanPrel findLeftScanPrel(String fieldName, RelNode leftRelNode) {
```
Due to the fact that HashJoin's buildSchema() method will prefetch both sides first batch data from both sides, if there's a blocking node under the probe side one , the corresponding probe side scan node will already run over before the build side bloom filter generated. So I will update that to still keep the blocking node checking logic.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterManager.java:
```diff
@@ -0,0 +1,586 @@
+      } else {
+        return null;
+      }
+    } else {
```
will update that.

weijietong commented:
@amansinha100 thanks for your valuable review. Since being on vacation , others will be commented and updated later.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/BloomFilter.java:
```diff
@@ -0,0 +1,176 @@
+  }
+
+  private void setMask(int key) {
+    final int SALT[] = {0x47b6137b, 0x44974d91, 0x8824ad5b, 0xa2b7289d, 0x705495c7, 0x2df1424b, 0x9efc4947, 0x5c6bfb31};
```
those 8 odd numbers was picked from Impala's implementation(see bloom-filter.h) . They act as salt values to participate in calculating the mask. I will add some comments to this method to make it clear.

amansinha100 replied:
Although I understand these are open source software but in general it is better to avoid referencing another project's implementation.  In this case, it is probably ok since picking these values is similar to picking prime numbers for hash computations and typically they are common across implementations.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/BloomFilter.java:
```diff
@@ -0,0 +1,176 @@
+  // Bytes in a bucket.
+  private static final int BYTES_PER_BUCKET = 32;
+  // Minimum bloom filter data size.
+  private static final int MINIMUM_BLOOM_SIZE = 256;
```
BYTES is right. will update that naming.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/BloomFilter.java:
```diff
@@ -0,0 +1,176 @@
+
+  private int mask[] = new int[8];
+
+  private byte[] tempBucket = new byte[32];
```
The reason to keep it as a member variable is to avoid heap memory GC while invoking the corresponding methods frequently. If move it to the locally methods, it will break down the performance .

amansinha100 replied:
Sounds good..but do you need to clear the tempBucket array across invocations ?

weijietong replied:
The `DrillBuf.getBytes(int index, byte[] dst)` will override the dst array's content. And we always invoke the getBytes method before usage the dst array. So there's no risk to use a polluted byte array.

weijietong commented:
To the cleanup point, I currently think there's no problem. Once the query was canceled, the foreman or the drillbit will run out , the in-flight runtime filter will failed due to the wire termination of one side.  The off-heap memory occupied by the bloom filter will be released by the promise of using `SendingAccountor` to send out `RuntimeFilter` through DataTunnel. If we find any potential issues later , we can fixed them at the following JIRAs.

weijietong commented:
@amansinha100 please review the changes made according to your advices.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java:
```diff
@@ -528,6 +590,27 @@ private void initializeBuild() {
+    }
+    runtimeFilterReporter = new RuntimeFilterReporter((ExecutorFragmentContext) context);
+    RuntimeFilterDef runtimeFilterDef = popConfig.getRuntimeFilterDef();
+    if (runtimeFilterDef != null) {
```
`enableRuntimeFilter` is a global option to decide whether a HashJoin could generate RuntimeFilter. But it does not indicate the HashJoin must have a RuntimeFilterDef. There's some preconditions defined at the `RuntimeFilterManager` such as Left join is not allowed to use a BF to filter data. So we can't use `Preconditions.checkState(runtimeFilterDef != null)`.

amansinha100 replied:
This would also depend on the NDV once we have proper stats.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/BloomFilter.java:
```diff
@@ -0,0 +1,176 @@
+  public BloomFilter(DrillBuf byteBuf) {
+    this.byteBuf = byteBuf;
+    this.numBytes = byteBuf.capacity();
+    this.byteBuf.writerIndex(numBytes);
```
the DrillBuf's content will be changed by calling setBytes and getBytes which not affect the readerIndex and writerIndex. When the BloomFilter was created, the DrillBuf's size is sure. The writerIndex setting here in advance is also clear and sure.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/BloomFilterCreator.java:
```diff
@@ -0,0 +1,41 @@
+import io.netty.buffer.DrillBuf;
+import org.apache.drill.exec.memory.BufferAllocator;
+
+public class BloomFilterCreator {
```
make sense.

amansinha100 commented:
Since most of my comments have been addressed,  I am +1.   Thanks @weijietong.  
Once you address @sohami 's comments,  the next steps would be: 
  - pls file JIRAs for items we identified during code review. 
 - rebase on master and run unit tests
 - mark the JIRA ready-to-commit
 - we'll need a volunteer to run the regression tests on QA test cluster (I will ask for volunteer)
 - presumably with the default FALSE setting for the JPPD feature, tests are clean, then it can be merged to master. 
 - continue further tests with a TRUE setting

weijietong commented:
@amansinha100 The following JIRAs: DRILL-6573 ,DRILL-6572 would solve what we identified before .  @sohami 's comments have been addressed.

amansinha100 commented:
@sohami pls see if your comments are addressed.  @weijietong just FYI, the ongoing 1.14.0 release is yet to be finished (it is in voting stage), so there may be some delays in  running regression tests on the QA cluster with your PR.

sohami commented:
Sorry for delayed response but I promise to complete review by this weekend.

sohami commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/WorkManager.java:
```diff
@@ -375,6 +378,30 @@ protected void cleanup() {
+      if (toForeman) {
+        Foreman foreman = queries.get(queryId);
+        if (foreman != null) {
+          foreman.getRuntimeFilterManager().registerRuntimeFilter(runtimeFilter);
```
the `registerRuntimeFilter` and eventual broadcast all is happening on rpc thread which is not good since Rpc thread should be released ASAP. Please follow a model similar to **cancelForeman** in WorkManager to submit the `registerRuntimeFilter` task as new Runnable to WorkManager executor

sohami commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/Foreman.java:
```diff
@@ -408,6 +424,9 @@ private void runPhysicalPlan(final PhysicalPlan plan) throws ExecutionSetupExcep
 
     startQueryProcessing();
+    if (enableRuntimeFilter) {
+      runtimeFilterManager.waitForComplete();
```
Why is this needed here ?

weijietong replied:
To make sure the `RuntimerFilterManager`'s sending out  `ByteBuf`s to be safely cleared out no matter the network successed or failed.

sohami replied:
But based on my understanding `RuntimeFilterManager` only sends out something over network during broadcasting of aggregated bloomfilter. Broadcast currently happens on the Rpc thread which needs to be changed. So at what time will this Foreman thread will end up also sending the buffers over network ?

weijietong replied:
There is various cases that the network between the foreman and receivers break down. Also the norm case is the receiver finished earlier before receiving the bloom filter. So the security way is to use the `SendingAccountor` to release the allocated `ByteBuf`.

sohami replied:
`SendingAccountor` on which this Foreman thread waits for completion is local to the `RunTimeFilterManager`. The job of `SendingAccountor` is to wait the caller thread iff there is any pending message to be sent unless the send is completed successfully or failed.  The buffer release is handled in the failure path of the mesaage not in sending accountor itself. The only time this `SendingAccountor` is incremented is when there is aggregated filter to be broadcasted. 
By the time this `Foreman` thread is exited there is no guarantee that remote minor fragments may have completed sending all their respective `Bloomfilters` and that `aggregated filter` is also broadcasted, hence it will be no-op in those cases. And more over if there is any failure case as you mentioned above there are chances of memory leak. To avoid that it will make more sense to put this `runtimeFilterManager.waitForComplete` inside `ForemanResult::close` method right before `closeFuture.removeListener(closeListener)`

weijietong replied:
sorry for the wrong state. will change it to the right position .

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterManager.java:
```diff
@@ -0,0 +1,666 @@
+      if (currentPrel instanceof HashAggPrel) {
+        encounteredBlockNode = true;
+        return null;
+      }
```
The blocking operators we care about here are those will wait for all the next batches to come in to output the result batches. So the `Sort`, `TopN` need to be considered. Thanks for noticing that, will update that. `Join` I think is not proper here . The `HashJoin`'s probe stage is not a blocking process. The `MergeJoin` is also not a blocking one.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterManager.java:
```diff
@@ -0,0 +1,666 @@
+          Wrapper container = findPhysicalOpContainer(rootWrapper, hashJoinPOP);
+          int majorFragmentId = container.getMajorFragmentId();
+          holder.setJoinMajorId(majorFragmentId);
+          Wrapper probeSideScanContainer = findPhysicalOpContainer(container, probeSideScanOp);
```
Could you explain why this will return null. To my understanding , once the code reaches here,  the HashJoinPoP is certain to own a `Wrapper`.

sohami replied:
Comment was based on `findPhysicalOpContainer` return values. If null will never be returned then we should modify the method `findPhysicalOpContainer` to not do so. In cases when nothing is found to return that's an IllegalState

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/ScanBatch.java:
```diff
@@ -190,11 +213,21 @@ public IterOutcome next() {
+            runtimeFiltered = true;
+          } else {
+            container.buildSchema(SelectionVectorMode.NONE);
+          }
```
I will take into account this suggestion.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/BloomFilter.java:
```diff
@@ -0,0 +1,190 @@
+    int thisLength = this.byteBuf.capacity();
+    Preconditions.checkArgument(otherLength == thisLength);
+    byte[] otherTmpBucket = new byte[BYTES_PER_BUCKET];
+    for (int i = 0; i < thisLength / BYTES_PER_BUCKET; i++) {
```
Yes, it's not possible to have not aligned buffers to aggregate.

sohami commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterManager.java:
```diff
@@ -0,0 +1,666 @@
+
+        @Override
+        public void interrupt(final InterruptedException e) {
+          //logger.error("fail to broadcast a runtime filter to the probe side scan node", e);
```
should we log this failure here and above ? Looks like if there is any failure in broadcasting aggregated filter we are treating it as minor event since query can still proceed without it.

weijietong replied:
As the runtime filter processing is async , the error log here is not so important. I am not sure to log it or not ,  so I comment  it here.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterManager.java:
```diff
@@ -0,0 +1,666 @@
+      if (aggregatedRuntimeFilter == null) {
+        aggregatedRuntimeFilter = runtimeFilterWritable;
+      } else {
+        this.aggregatedRuntimeFilter.aggregate(runtimeFilterWritable);
```
`aggregatedRuntimeFilter` is unique to a majorId.  But still has bugs here ,I have updated to have a majorId-aggregatedRuntimeFilter map. Thanks.

weijietong commented:
@sohami I have changed the logic to generate SV2 in the ScanBatch. It is comparatively easy now. 

Initially, I worried about the performance of the `RemovingRecordBatch`'s `setupNewSchema()` while generating different output schema of `ScanBatch` by applying `RuntimeFilter` with an assumption implementation of the `RemovingRecodBatch`'s `setupNewSchema()` to generate dynamic codes.But actually `RemovingRecodBatch`'s `setupNewSchema()` does not generate dynamic codes. So the changed implementation has reduced some complexity to judge whether to output a schema with SV2 or NONE.

I am not sure I fully understand your proposed strategy , maybe a implementation like adding `RemovingRecodBatch` on a node . But the whole design is to connect the `Scan` node and `HashJoin` one. That changing will make more work to do. So if you have no doubt about the current changed  implementation , let's it go.

Btw, others you mentioned has been addressed. Please review them, tks!

sohami commented:
Overall other changes looks good but `ScanBatch` still has issues as before. I will comment in JIRA for ScanBatch Issue since github is very slow. Also for future reference please add separate commit for changes made to address review comments, that will make it easier to review just those changes. :)

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterManager.java:
```diff
@@ -0,0 +1,677 @@
+    }
+    List<Wrapper> dependencies = wrapper.getFragmentDependencies();
+    if (CollectionUtils.isEmpty(dependencies)) {
+      return null;
```
No ,it's a right case to a join node with one side dependency having container,one side dependency not having .

weijietong commented:
ScanBatch’s discussion is on JIRA,please see that @sohami.

weijietong commented:
@sohami  this commit has added a `RuntimeFilterRecordBath` to replace the related work of the `ScanBatch`. Please review, thanks.

sohami commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/filter/RuntimeFilterRecordBatch.java:
```diff
@@ -0,0 +1,222 @@
+      } catch (Exception e) {
+        throw UserException.internalError(e).build(logger);
+      }
+    }
```
above setup of `Hash64` should not happen for each `incoming` record batch, since that will be very expensive. It should happen only once during `setupNewSchema` phase until next `OK_NEW_SCHEMA`.

weijietong replied:
When the `setupNewSchema` was called, the `RuntimeFilterWritable` may have not been received . So the `hash64` should be initialized at the `doWork`. Since there's `hash64 == null` checking and initializing `hash64` logic, this part of code will only run one time.

weijietong replied:
make sense, but could you explain at which case ,to a query , which will have consecutive incoming `RecordBatch` but some with different schema? This really confuse me and make the solving logic complex.

sohami replied:
a simple example will be Scan is reading 2 json files in 2 different record batches. Both files have a column of type map lets say m. File1 have children in map a,b of type int whereas file2 have children a,b,c. In this case scan will generate OK_NEW_SCHEMA

sohami commented:
@weijietong - Thanks for making the changes. But I am seeing issues in implementation related to how internals work, following protocol of iterative model of Drill and few other things. I have left few comments for now and was wondering that it will be easier to provide a commit myself to address those issues and help you get this PR pushed in faster.

Also I was working on a change for SV2 optimization [here](https://github.com/sohami/drill/commits/SV2_Optimization). Given you also have subset of the change for it with some issues, it would be great if you can rebase on top of my change.
Please let me know what you think.

weijietong commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/filter/RuntimeFilterRecordBatch.java:
```diff
@@ -0,0 +1,222 @@
+  private void applyRuntimeFilter() throws SchemaChangeException {
+    RuntimeFilterWritable runtimeFilterWritable = context.getRuntimeFilter();
+    if (runtimeFilterWritable == null) {
+      sv2.setRecordCount(incoming.getRecordCount());
```
Since the refactored  `GenericSV2Copier` treat the `originalRecodCount == recordCount` case special. There's no need to do this addition work.

weijietong replied:
make sense ,will update.

weijietong commented:
@sohami Since we move the filter logic from `ScanBatch` into the `RuntimeFilterRecordBatch` , I refactor the `RuntimeFilterManager` to move the JPPD judgement logic into `RuntimeFilterVisitor`. These part of codes have fewer changes than before.  

Your SV2 optimization is more general than mine. I don't know how to operate to rebase your own branch into mine. Maybe you can guide me or I fire another JIRA to refactor the implementation to yours one after this PR and yours were merged.

weijietong commented:
Btw, to your optimization, I suggest the initial value `batchActualRecordCount` of `SelectionVector2` to be -1 ,not a positive number.

parthchandra commented on protocol/src/main/protobuf/UserBitShared.proto:
```diff
@@ -343,6 +343,7 @@ enum CoreOperatorType {
   IMAGE_SUB_SCAN = 52;
   SEQUENCE_SUB_SCAN = 53;
   PARTITION_LIMIT = 54;
+  RUNTIME_FILTER = 55;
```
It would be OK to not generate c++ protobufs in this PR since the C++ client doesn't actually (need to) use these values. We can lump the protobuf regeneration with some other C++ client PR.

weijietong commented:
@sohami @parthchandra  thanks for the help. I have addressed the comments according to your branch. Btw, I suggest the SV2's `batchActualRecordCount` be renamed as `originalRecordCount`.  The `recordCount` member is the truly final actual record count. Otherwise ,this naming will make others confused. If you agree with it, I will update that.

sohami commented:
@weijietong - Would have been easier to just cherry-pick the commits on top of your rebased branch. But I guess it's fine. 
`recordCount` is not the actual record count, its the selected number of records. I have also added comments to differentiate between the two variables.

Few things to resolve:
1) Since I have already shared a commit with generated protobuf files for c++ client. It would be good to take that as well.
2) Please resolve the travis build failure.
3) Meanwhile I ran tests on my shared branch yesterday and there are unit tests failures. Can you run unit tests locally and make sure all of them are passing.

weijietong commented:
@sohami tests have passed by removing the Kind.Integer from the OptionValue. Still keep the `IntegerValidator` to validate the number which should be a integer range value.

sohami commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/options/TypeValidators.java:
```diff
@@ -156,9 +156,19 @@ public DoubleValidator(String name) {
+      super.validate(v, metaData, manager);
+      if (v.num_val > Integer.MAX_VALUE || v.num_val < Integer.MIN_VALUE) {
+        throw UserException.validationError()
+          .message(String.format("Option %s does not have a validate integer value", getOptionName()))
```
does not have a **valid** integer value

sohami commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/svremover/AbstractSV2Copier.java:
```diff
@@ -32,6 +32,7 @@
   private SelectionVector2 sv2;
   protected List<TransferPair> transferPairs = new ArrayList<>();
 
+
```
please revert back this change.

sohami commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/svremover/RemovingRecordBatch.java:
```diff
@@ -27,6 +27,7 @@
 import org.apache.drill.exec.record.VectorWrapper;
 import org.apache.drill.exec.record.WritableBatch;
 
+
```
please revert back this extra line here and below.

sohami commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/svremover/StraightCopier.java:
```diff
@@ -26,6 +26,7 @@
 
 import java.util.List;
 
+
```
please revert back extra line

sohami commented on exec/java-exec/src/main/java/org/apache/drill/exec/record/selection/SelectionVector2.java:
```diff
@@ -179,6 +179,7 @@ public void close() {
     clear();
   }
 
+
```
remove this extra line

sohami commented on pom.xml:
```diff
@@ -389,6 +389,7 @@
         <groupId>com.mycila</groupId>
         <artifactId>license-maven-plugin</artifactId>
       </plugin>
+
```
revert this change.

sohami commented:
Thanks for all the changes. I ran the regression tests and everything was green. 
+1 LGTM from my side.

amansinha100 commented:
@sohami thanks for the detailed follow-up review and running the regression tests with the PR (I assume with default setting of run-time pushdown disabled).  @weijietong sorry I have not been responsive lately due to personal reasons.   Thanks for addressing review comments. 
@sohami , it's a +1 from me as well, so when you get a chance pls go ahead and merge the PR.

arina-ielchiieva commented:
@weijietong please rebase and resolve the conflicts (would you need to regenerate the protobufs).

arina-ielchiieva commented:
@weijietong thanks!

arina-ielchiieva commented:
@weijietong run the tests on your branch and have noticed that

 though all unit tests pass there are a huge amount of warnings during the run, checked on master such warnings are absent. Checked a couple of times and got consistent results. Attaching output of two runs for comparison. You may see the difference even in files sizes (2.53 MB vs 6.45 MB).

[consoleText_master.txt](https://github.com/apache/drill/files/2310225/consoleText_master.txt)
[consoleText_jppd.txt](https://github.com/apache/drill/files/2310229/consoleText_jppd.txt)

Please take a look and fix.

sohami commented:
@arina-ielchiieva  - Nice catch.
@weijietong  - Please remove [this line](https://github.com/apache/drill/pull/1334/files#diff-e58119193708a202639717568263672fR43) and line 78 from same file. The extra WARN lines are because of enabling logger in those tests cases. Since logger is not closed it's not restored back, hence for all future tests WARN level log is enabled.

weijietong commented:
@arina-ielchiieva @sohami  thanks for the detail catch, done.
